crates/ra_ide_api/src/completion/completion_item.rs

```diff
@@ -78,72 +110,111 @@ impl CompletionItem {
             .map(|it| it.as_str())
             .unwrap_or(self.label())
     }
-    /// What is inserted.
-    pub fn insert_text(&self) -> InsertText {
-        match &self.snippet {
-            None => InsertText::PlainText {
-                text: self.label.clone(),
-            },
-            Some(it) => InsertText::Snippet { text: it.clone() },
-        }
+
+    pub fn insert_text_format(&self) -> InsertTextFormat {
+        self.insert_text_format.clone()
     }
 
     pub fn kind(&self) -> Option<CompletionItemKind> {
         self.kind
     }
+    pub fn text_edit(&mut self) -> Option<&AtomTextEdit> {
+        self.text_edit.as_ref()
+    }
+    pub fn take_additional_text_edits(&mut self) -> Option<TextEdit> {
+        self.additional_text_edits.take()
+    }
 }
 
 /// A helper to make `CompletionItem`s.
 #[must_use]
-pub(crate) struct Builder {
+pub(crate) struct Builder<'a> {
+    ctx: &'a CompletionContext<'a>,
```

Comment:
It feels too coarse-grained to add the whole `CompletionContext` to the `Builder`. Could we add a `delete_range: TextRange` argument to `new` instead?

Reply:
we need both delete_range and offset. should we add both of them to `new`?

Reply:
leaf_range is enough. my old code use `offset`. i refactored it once.